### DISCUSSION TITLE
Support for images

### DISCUSSION THREAD

**Discussion by dberenbaum** [treeverse/dvc#5681] [New Features & Ideas]
Image files can be important in dvc projects because:
* It's often easier to make custom plots as image files rather than using dvc plots.
* Especially in fields like computer vision, there may be other image types that need to be compared between commits.

I initially thought this was a straightforward feature to implement, but I think there are questions to discuss:

1. Should there be dvc-specific support for image diffs? What should the UI and output look like (for example, `dvc images diff`)? Should it support comparison of two images like `metrics diff` or more than two like `plots diff`?
2. Does dvc need a way to specify whether outputs are images (for example, adding an `image: true` option under a `dvc.yaml` output)? This might be helpful to show/diff all images in a project, and to enable support for apps like viewer to show all images. Could image files instead be inferred from file extensions?

Some examples of diffing images:
https://docs.github.com/en/github/managing-files-in-a-repository/rendering-and-diffing-images
https://github.com/ewanmellor/git-diff-image
https://oscarnajera.com/2017/11/git-diff-images-and-pdfs/

@pared @mnrozhkov @JIoJIaJIu @dmpetrov

**Comment by pared** [comment]
> Does dvc need a way to specify whether outputs are images 

Currently, we are determining the file type by extension. I think we could do the same for images. Let's start with supporting some popular formats, and implement new ones per users requests.

> Should there be dvc-specific support for image diffs? What should the UI and output look like (for example, dvc images diff)? Should it support comparison of two images like metrics diff or more than two like plots diff?

I guess that depends on what we are actually expecting from `dvc plots show` for images. If we do expect it to embed the target image into HTML, then we could probably solve the `diff` problem by just showing the targets from different branches one near another, just like we do with the confusion matrix template in the current version of plots.

**Reply by skshetry** [reply]
> Currently, we are determining the file type by extension. I think we could do the same for images. Let's start with supporting some popular formats, and implement new ones per users requests.

We could also try to detect if they are image using image headers and then fallback to using filename or just assume it to be a non-image. 
https://docs.python.org/3/library/imghdr.html
https://github.com/h2non/filetype.py

P.S: This does mean that we'll have to read first few bytes, so it's not fast and easy to determine compared to the extension which we'll always have on hand, even if dvc doesn't have a file on hand.

**Comment by dberenbaum** [comment]
After thinking about it, I lean towards forcing users to be explicit about images. Thinking about a use case like computer vision, there may be lots of image files in the raw data, and then a small number of images used for reporting, plots, etc. It might be important to have dvc only manage the specified images.

> I guess that depends on what we are actually expecting from `dvc plots show` for images.

Do you see this as being part of `dvc plots show`? I was thinking it was more likely to be a different command, like maybe `dvc images show/diff`.

**Reply by pared** [reply]
> Do you see this as being part of dvc plots show? I was thinking it was more likely to be a different command, like maybe dvc images show/diff.

That might be a better idea. The question is, do we want to expand, already big, amount of commands. Maybe we should pack  them into a subcommand? Like `dvc vis plots` and `dvc vis images`?

**Reply by dberenbaum** [reply]
That's a good idea, and maybe even combine metrics (and params?) into the same subcommand. I'm thinking through proposals to unify the various show/diff commands.

Introducing the subcommand would be a breaking change, though, right? Maybe it's better to implement as a separate command first and then reorganize commands later.

**Reply by pared** [reply]
Sure, reorganization should probably go for next major release.

**Comment by pmrowla** [comment]
I don't really see why this needs to be core DVC functionality. DVC should try to be data-type agnostic, and I'm not sure that setting a precedent for handling specific file types is a good idea.

To me, this sounds like a use case where a user should be using `api.open()`/`api.read()` to get file handles for the 2 (or more) versions of a file they want to binary diff, and then feed that data into their preferred diffing tool. (In this case, any one of the widely available image processing/diffing tools that are out there, like imagemagick or PIL/Pillow)

**Reply by pmrowla** [reply]
A standalone CLI tool to do this that depends on DVC + Pillow would be simple to implement, and maybe even a good "example DVC API usage" project, but I don't think it belongs as a part of core DVC.

**Reply by pared** [reply]
The main reason to start supporting images is to not to make users use our templates, if they already have a project producing some visualizations. I agree that it might be a pandora box that may flood us with requests for different visualizations types. But even if we do not decide to implement support for images, we should probably make the mentioned `DVC + PIllow` example.

**Reply by dberenbaum** [reply]
> I don't really see why this needs to be core DVC functionality. DVC should try to be data-type agnostic, and I'm not sure that setting a precedent for handling specific file types is a good idea.

As a counterpoint, how is this different from `dvc plots`?

We should distinguish between 1) the UI of showing image diffs and 2) identifying different data types. I agree that DVC itself should have limited UI support, but I'd argue that DVC is already not data-type agnostic.

Let's say a user is doing a computer vision problem, like the cats and dogs one in the dvc tutorial: https://dvc.org/doc/use-cases/versioning-data-and-model-files/tutorial. They may have thousands (or many more) of images. They may also have some summary image outputs, which could be plots or ways to visualize their networks/weights. It would be great for downstream apps like Studio to know which images that user wants to show/diff. Where should the user identify those image outputs so that front-end apps can display them?

**Reply by pmrowla** [reply]
> As a counterpoint, how is this different from dvc plots?

I would say that doing too much with plots is also something we should not do in core DVC :)

We provide the basic functionality for visualizing structured metrics (and metrics must follow a format that we have control over and enforce), and some basic templates. Beyond that, if users want more advanced visualizations they can define their own plot templates using vega, or just use DVC to retrieve the raw metrics data and then do whatever they need to with it.

IMO we should approach binary diffing in the same way - provide an API for the user to get the requested versions of the data they want, and then leave it up to the user what to do with that data.

**Reply by pmrowla** [reply]
> Where should the user identify those image outputs so that front-end apps can display them?

If the issue is just "how does the frontend app know which DVC-tracked files can be displayed visually", a simple solution would be for us to define an optional mimetype metadata/schema field for outputs. If a user knows a file is going to be an image (or a directory which only contains images), they could tag it with the appropriate mimetype themselves in `dvc.yaml`. Core DVC would not do anything with this field, but downstream UI tools like studio, (or a user's own hypothetical image diffing utility) could do whatever they need to with that information.

**Reply by pared** [reply]
Maybe what do we need is a way for users to `api.get_url` but in an aggregated way? 
What I mean, is that we don't know what user will want to do with the binary target for diff - it can be image, it can be audio file, whatever.
We could however provide a method that would allow users to get all the required URLs from remote/cache. For example, user could ask "give me file_x and file_y for revisions rev1, rev2, rev3" and API could return structured data so that user can process that later in a way is suitable for the specific use case.

**Reply by dberenbaum** [reply]
@pmrowla That makes sense to me, and I think we are mostly in agreement? I tried to address some of these issues in the proposal for a unified diff (https://github.com/iterative/enhancement-proposals/pull/4). We can continue discussion in either place.

@pared That seems pretty close to the diff driver suggestion in https://github.com/iterative/enhancement-proposals/pull/4. There might need to be some minimal additional parsing for each type of output, like converting all metrics file types (json, yaml, etc.) to one consistent data schema.

**Reply by pared** [reply]
@dberenbaum I agree that it looks like a driver specification. Though I would argue that this use case is a little bit different: when we are talking about iterative/enhancement-proposals#4 we talk about `diff` command. Here, I am thinking about including some abstract driver into `dvc.api` so that users could handle their specific use cases by themselves if `metrics/plots/params(?)/images(?)` don't fill their needs.

**Reply by dberenbaum** [reply]
@pared Maybe I'm missing your point. I agree that it makes sense to have an abstract driver to handle custom needs outside of `metrics/plots/params(?)/images(?)`.  What is special about those other than support for `diff` (and `show`)?

**Reply by pared** [reply]
Well I guess the main point here is to let user implement own driver. I imagined it as a code-related task, and something user would be handling in his environment rather than cli but looking at the discussion [under `diff` issue](https://github.com/iterative/enhancement-proposals/pull/4#issuecomment-810688518), my view on user- implemented driver is what you call difftool - so it seems to me we agree in that matter.

**Comment by pared** [comment]
It seems that subject raised in this discussion is also discussed in https://github.com/iterative/enhancement-proposals/pull/4

**Comment by mnrozhkov** [comment]
Thanks for starting this discussion! Let me share my thoughts and expectations about this feature. 

First, short thoughts on discussion above
1 - dvc-specific support for image diffs like `dvc images diff` seems unnecessary (images is just one type of content DVC may visualize with `dvc plots`) and it's enough to embed them into current plots functionality 
2 - for me, the most natural way to work with images - `dvc.yaml` where I would specify some plot as an image   
```yaml 
train_model:
 cmd: python train_model.py --base_config=config/params.yaml
 deps:
   - traintest.csv
   - train_model.py
 params:
   - config/params.yaml:
       - base
       - train
 outs:
   - model.pickle
 plots:
   - reports/grid_search.png:
       cache: true
       image: true
       title: Grid Search results
```
3 - with such `dvc.yaml` I could run `dvc plots show/diff` and get my images in generated `plots.html` file  (let me share this slide below)
![image](https://user-images.githubusercontent.com/11489539/113839544-71807800-9798-11eb-832d-8714b93576d5.png)

4 - I don't know about use case for computer vision though.

**Reply by dberenbaum** [reply]
Thanks, @mnrozhkov! I think this makes sense for your use case, but I worry that it may not generalize well for a few reasons:

- Other users may want to do the same for images that are not plots (see the image below for an example of what a computer vision diff might look like). In that case, would it be a separate section in `dvc.yaml`, or should plots still be used?
- In terms of implementation, the currently supported plots and the images would be quite different, so it might be hard to combine them. For example, the plots have a template option that wouldn't make much sense for images.
- DVC is primarily a command-line tool, and plots are already an anomaly in that they generate HTML output. It might make more sense for the slides you show above to be part of Studio or some other front-end extension and for DVC to only manage the data to generate those visualizations.

![image](https://user-images.githubusercontent.com/2308172/113925071-91c02f00-97b8-11eb-8514-c11bf5f97667.png)

**Comment by dberenbaum** [comment]
How should image plots be identified?

Here are a few proposals (feel free to suggest others):

1. They must all be added explicitly to `dvc.yaml` (see the example above from @mnrozhkov).
2. Any file added to a specified plots or images directory path will be read like a plot regardless of whether it's in `dvc.yaml` at all.
3. Some kind of automatic heuristic can be used (any file with the appropriate extension stored in the Git repo).

These could also be combined so that there is a default behavior but also ways to be more flexible (similar to params reading automatically from `params.yaml` but also allowing custom params files).
